source/extensions/filters/network/postgres_proxy/postgres_decoder.cc

```diff
@@ -198,44 +198,65 @@ bool DecoderImpl::parseHeader(Buffer::Instance& data) {
     ENVOY_LOG(trace, "postgres_proxy: cannot parse message. Need {} bytes in buffer",
               message_len_ + (startup_ ? 0 : 1));
     // Not enough data in the buffer.
-    return false;
+    return Decoder::NeedMoreData;
   }
 
   if (startup_) {
     uint32_t code = data.peekBEInt<uint32_t>(4);
     // Startup message with 1234 in the most significant 16 bits
     // indicate request to encrypt.
     if (code >= 0x04d20000) {
-      ENVOY_LOG(trace, "postgres_proxy: detected encrypted traffic.");
       encrypted_ = true;
-      startup_ = false;
-      incSessionsEncrypted();
+      // Handler for SSLRequest (Int32(80877103) = 0x04d2162f)
+      // See details in https://www.postgresql.org/docs/current/protocol-message-formats.html.
+      if (code == 0x04d2162f) {
+        // Notify the filter that `SSLRequest` message was decoded.
+        // If the filter returns true, it means to pass the message upstream
+        // to the server. If it returns false it means, that filter will try
+        // to terminate SSL session and SSLRequest should not be passed to the
+        // server.
+        ENVOY_LOG(trace, "postgres_proxy: enabling secure transport.");
+        encrypted_ = callbacks_->onSSLRequest();
```

Comment:
Should we log the returned value as well?

Reply:
I removed the log from here and added it to `onSSLRequest` method. When terminateSSL option is enabled, log will indicate if it was successful or not. If terminateSSL is disabled, nothing will be logged and messages will pass in encrypted form without decoding.

Reply:
Cool. Thanks!